# Ravjot Sachdev
## CIS 700 Midterm
## Topic 1: Evading Malware Detection
## May 23, 2021

MalGAN code was downloaded from https://github.com/ZaydH/MalwareGAN. One slight modification was made to allow for Windows compatability.

In [1]:
%cd MalwareGAN

C:\Users\ravsa\Downloads\SU\Spring 2021\Midterm_Spring_2021\MalwareGAN


First, we install the necessary packages to run the code.

In [2]:
!pip install -r "requirements.txt"

We can run the code with the provided trial data. The first file contains the feature vector of a single malicious file, while the second file contains the feature vector of a benign file. For this scenario, we'll use the recommended batch size of 32 and number of epochs of 100.

In [3]:
!python main.py 10 32 100 data/trial_mal.npy data/trial_ben.npy

Filename =  logs_2021-05-23-20-59-59.716294.log
2021-05-23 20:59:59,717 -- INFO -- ******************* New Run Beginning *****************
2021-05-23 20:59:59,717 -- DEBUG -- CUDA: Disabled
2021-05-23 20:59:59,717 -- INFO -- main.py 10 32 100 data/trial_mal.npy data/trial_ben.npy
2021-05-23 20:59:59,725 -- DEBUG -- Constructing new MalGAN
2021-05-23 20:59:59,725 -- DEBUG -- Malware Dimension (M): 128
2021-05-23 20:59:59,726 -- DEBUG -- Latent Dimension (Z): 10
2021-05-23 20:59:59,726 -- DEBUG -- Test Split Ratio: 0.200
2021-05-23 20:59:59,726 -- DEBUG -- Generator Hidden Layer Sizes: [256, 256]
2021-05-23 20:59:59,726 -- DEBUG -- Discriminator Hidden Layer Sizes: [256, 256]
2021-05-23 20:59:59,727 -- DEBUG -- Blackbox Detector Type: RandomForest
2021-05-23 20:59:59,727 -- DEBUG -- Activation Type: LeakyReLU
2021-05-23 20:59:59,731 -- DEBUG -- Starting training of blackbox detector of type "RandomForest"
2021-05-23 20:59:59,874 -- DEBUG -- COMPLETED training of blackbox detector of type

Traceback (most recent call last):
  File "main.py", line 152, in <module>
    main()
  File "main.py", line 146, in main
    results = malgan.measure_and_export_results()
  File "C:\Users\ravsa\Downloads\SU\Spring 2021\Midterm_Spring_2021\MalwareGAN\malgan\__init__.py", line 409, in measure_and_export_results
    y_mal_orig, y_prob, y_hat_post)
  File "C:\Users\ravsa\Downloads\SU\Spring 2021\Midterm_Spring_2021\MalwareGAN\malgan\_export_results.py", line 38, in _export_results
    with open(results_file, "a+") as f_out:
PermissionError: [Errno 13] Permission denied: 'results.csv'


In [4]:
!pip install pandas
import pandas as pd
results = pd.read_csv("results.csv")
results.head()

  Using cached numpy-1.20.3-cp37-cp37m-win_amd64.whl (13.6 MB)
  Attempting uninstall: numpy
    Found existing installation: numpy 1.16.1
    Uninstalling numpy-1.16.1:
      Successfully uninstalled numpy-1.16.1


time_completed    M   Z  batch_size  test_set_size  \
0  2021-05-23 20:45:10.090726  128  10          32            361   
1  2021-05-23 20:48:47.292262  128  10          32            361   

  detector_type activation gen_hidden_dim discim_hidden_dim  \
0  RandomForest  LeakyReLU     [256, 256]        [256, 256]   
1  RandomForest  LeakyReLU     [256, 256]        [256, 256]   

   avg_validation_loss  avg_test_loss  avg_num_bits_changed       auc  \
0            -1.146592      -1.178672                25.322  0.902160   
1            -1.139809      -1.039155                24.985  0.940185   

   orig_mal_detect_rate  mod_mal_detect_rate  ben_mal_detect_rate  
0              0.981685             0.604396             0.102273  
1              0.989011             0.604396             0.090909

## Results
The CSV file above contains the relevant results. As seen in the final columns, the original detection rate of the malware classifier was above 98%. After running MalGAN and generating adversarial samples, the detection rate fell to just 9%. This shows MalGAN's effectiveness at modifying malicious program to bypass malware detection neural networks. While MalGAN is not the only machine learning model to achieve this goal, its improvements on previous models makes it a great choice to attack such malware detectors.